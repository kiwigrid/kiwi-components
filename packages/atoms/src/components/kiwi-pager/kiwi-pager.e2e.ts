import { newE2EPage } from '@stencil/core/testing';

describe('kiwi-pager', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<kiwi-pager></kiwi-pager>');

    const element = await page.find('kiwi-pager');
    expect(element).toHaveClass('hydrated');
  });
});
