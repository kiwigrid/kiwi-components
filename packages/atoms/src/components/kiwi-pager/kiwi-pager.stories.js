import { html } from 'lit-html';
import notes from './readme.md';

export default {
  title: 'KiwiComponents/Pager',
  parameters: { docs: { description: { component: notes } } },
  decorators: [],
};

export const basic = (args) => {
  const { debounce, total, ofLabel } = args;

  return html` <div class="panel panel-default">
    <div class="panel-body">
      <kiwi-pager
        debounce=${debounce}
        page="0"
        total=${total}
        of-label=${ofLabel}
      />
    </div>
  </div>`;
};
basic.args = {
  debounce: 400,
  total: 5,
  ofLabel: 'of',
};
