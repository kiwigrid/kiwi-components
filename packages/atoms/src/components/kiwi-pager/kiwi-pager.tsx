import {
  Component,
  ComponentInterface,
  Event,
  EventEmitter,
  h,
  JSX,
  Prop,
  State,
} from '@stencil/core';
import { concat, Subject } from 'rxjs';
import {
  debounceTime,
  distinctUntilChanged,
  repeat,
  takeUntil,
} from 'rxjs/operators';

@Component({
  tag: 'kiwi-pager',
  shadow: false,
})
export class KiwiPager implements ComponentInterface {
  /**
   * Debounce time in milliseconds.
   */
  @Prop()
  debounce = 400;

  /**
   * Zero based page number
   */
  @Prop({ reflect: true, mutable: true })
  page!: number;

  /**
   * Total pages
   */
  @Prop({ reflect: true })
  total!: number;

  /**
   * Translated "of" label (e.g. 3 of 10)
   */
  @Prop()
  ofLabel!: string;

  @State()
  error = false;

  /**
   * Event signaling a change of the page number.
   */
  @Event({
    eventName: 'pageChanged',
    bubbles: true,
  })
  pageChanged!: EventEmitter<{ page: number }>;

  private input?: HTMLInputElement;
  private inputChangedDebouncer: Subject<number> = new Subject();
  private buttonChangedDebouncer: Subject<number> = new Subject();
  private cancelPageChange: Subject<void> = new Subject();

  private disconnect$: Subject<void> = new Subject();

  componentDidLoad(): void {
    concat(
      this.inputChangedDebouncer.pipe(
        distinctUntilChanged(),
        debounceTime(this.debounce),
        takeUntil(this.cancelPageChange),
        // repeats on completion (in case it got canceled by `cancelPageChange`)
        repeat(),
      ),
      this.buttonChangedDebouncer.pipe(debounceTime(200)),
    )
      .pipe(takeUntil(this.disconnect$))
      .subscribe((value: number) => {
        this.pageChanged.emit({ page: value });
      });
  }

  /* istanbul ignore next */
  disconnectedCallback(): void {
    this.disconnect$.next();
    this.disconnect$.complete();
  }

  render(): JSX.Element | null {
    if (this.total < 2) {
      return null;
    }

    return (
      <div class="pager" style={{ 'max-width': '250px' }}>
        <nav
          role="navigation"
          aria-label="Pagination Navigation"
          class="input-group"
        >
          <div class="pager-buttons input-group-btn">
            <button
              aria-label="Previous Page"
              type="button"
              class="btn btn-default"
              disabled={this.page === 0}
              onClick={this.handleDecrementPage}
            >
              <span class="glyphicon glyphicon-chevron-left" />
            </button>
          </div>
          <div class={this.error ? 'has-error' : ''}>
            <input
              min={1}
              max={this.total}
              type="text"
              class="form-control pager-input"
              value={this.page + 1}
              ref={(element) => {
                this.input = element;
              }}
              onInput={this.handleInputChange}
              arial-label="Current Page"
            />
          </div>
          <span class="input-group-addon">
            {this.ofLabel} {this.total}
          </span>
          <div class="pager-buttons input-group-btn">
            <button
              aria-label="Next Page"
              type="button"
              class="btn btn-default"
              disabled={this.page === this.total - 1}
              onClick={this.handleIncrementPage}
            >
              <span class="glyphicon glyphicon-chevron-right" />
            </button>
          </div>
        </nav>
      </div>
    );
  }

  private handleIncrementPage: () => void = () =>
    this.changePage(
      Math.min(this.page + 1, this.total - 1),
      this.buttonChangedDebouncer,
    );

  private handleDecrementPage: () => void = () =>
    this.changePage(Math.max(this.page - 1, 0), this.buttonChangedDebouncer);

  private handleInputChange: () => void = () => {
    try {
      const value: number = parseInt(this.input?.value ?? '') - 1;

      if (value >= this.total || value < 0 || isNaN(value)) {
        this.error = true;
        this.cancelPageChange.next();
        throw new Error('Validation failed');
      } else {
        this.changePage(value, this.inputChangedDebouncer);
      }
    } catch (e) {}
  };

  private changePage(number: number, debouncer: Subject<number>): void {
    this.error = false;
    this.page = number;
    debouncer.next(number);
  }
}
