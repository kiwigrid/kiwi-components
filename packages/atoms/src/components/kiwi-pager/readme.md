# kiwi-pager



<!-- Auto Generated Below -->


## Properties

| Property               | Attribute  | Description                                  | Type     | Default     |
| ---------------------- | ---------- | -------------------------------------------- | -------- | ----------- |
| `debounce`             | `debounce` | Debounce time in milliseconds (Default: 400) | `number` | `400`       |
| `ofLabel` _(required)_ | `of-label` | Translated "of" label (e.g. 3 of 10)         | `string` | `undefined` |
| `page` _(required)_    | `page`     | Zero based page number                       | `number` | `undefined` |
| `total` _(required)_   | `total`    | Total pages                                  | `number` | `undefined` |


## Events

| Event         | Description                                  | Type                             |
| ------------- | -------------------------------------------- | -------------------------------- |
| `pageChanged` | Event signaling a change of the page number. | `CustomEvent<{ page: number; }>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
