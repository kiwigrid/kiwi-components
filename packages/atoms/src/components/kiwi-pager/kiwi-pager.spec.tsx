import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { fakeSchedulers } from 'rxjs-marbles/jest';
import { expectDefined } from '../util/testing';
import { KiwiPager } from './kiwi-pager';

describe('kiwi-pager', () => {
  it('renders', async () => {
    const pager = await newSpecPage({
      components: [KiwiPager],
      template: () => (
        <kiwi-pager page={1} total={3} ofLabel="vong"></kiwi-pager>
      ),
    });

    expect(pager.root).toMatchSnapshot();

    pager.rootInstance?.disconnectedCallback();
  });

  it('renders empty', async () => {
    const pager = await newSpecPage({
      components: [KiwiPager],
      template: () => (
        <kiwi-pager page={0} total={0} ofLabel="vong"></kiwi-pager>
      ),
    });

    expect(pager.root).toMatchSnapshot();

    pager.rootInstance?.disconnectedCallback();
  });

  it('is does not exceed bounds', async () => {
    const pager = await newSpecPage({
      components: [KiwiPager],
      template: () => (
        <kiwi-pager page={0} total={3} ofLabel="vong"></kiwi-pager>
      ),
    });

    const prev = document.querySelector<HTMLButtonElement>(
      'button[aria-label="Previous Page"]',
    );
    const next = document.querySelector<HTMLButtonElement>(
      'button[aria-label="Next Page"]',
    );

    prev?.click();

    expect(pager.rootInstance.page).toEqual(0);

    next?.click();
    next?.click();

    expect(pager.rootInstance.page).toEqual(2);

    next?.click();

    expect(pager.rootInstance.page).toEqual(2);

    const input = document.querySelector<HTMLInputElement>('input.pager-input');

    expectDefined(input);

    input.value = '2';
    input.dispatchEvent(new Event('input'));

    expect(pager.rootInstance.page).toEqual(1);

    input.value = '5';
    input.dispatchEvent(new Event('input'));

    expect(pager.rootInstance.page).toEqual(1);

    input.value = '-5';
    input.dispatchEvent(new Event('input'));

    expect(pager.rootInstance.page).toEqual(1);

    input.value = 'asdf';
    input.dispatchEvent(new Event('input'));

    expect(pager.rootInstance.page).toEqual(1);

    pager.rootInstance?.disconnectedCallback();
  });

  it('emits pageChanged event after debounce', async () => {
    jest.useFakeTimers();

    const pager = await newSpecPage({
      components: [KiwiPager],
      template: () => (
        <kiwi-pager
          page={0}
          total={3}
          ofLabel="vong"
          debounce={400}
        ></kiwi-pager>
      ),
    });

    fakeSchedulers((advance) => {
      const pagerInst: KiwiPager = pager.rootInstance;

      jest.spyOn(pagerInst.pageChanged, 'emit');

      const input = document.querySelector<HTMLInputElement>(
        'input.pager-input',
      );

      expectDefined(input);

      input.value = '1';
      input.dispatchEvent(new Event('input'));
      input.value = '2';
      input.dispatchEvent(new Event('input'));
      input.value = '3';
      input.dispatchEvent(new Event('input'));

      advance(400);

      expect(pagerInst.pageChanged.emit).toBeCalledTimes(1);
      expect(pagerInst.pageChanged.emit).toBeCalledWith({ page: 2 });

      pager.rootInstance?.disconnectedCallback();
    })();

    jest.useRealTimers();
  });
});
