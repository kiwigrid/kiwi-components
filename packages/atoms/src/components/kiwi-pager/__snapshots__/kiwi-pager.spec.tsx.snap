// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kiwi-pager renders 1`] = `
<kiwi-pager page="1" total="3">
  <div class="pager" style="max-width: 250px;">
    <nav aria-label="Pagination Navigation" class="input-group" role="navigation">
      <div class="input-group-btn pager-buttons">
        <button aria-label="Previous Page" class="btn btn-default" type="button">
          <span class="glyphicon glyphicon-chevron-left"></span>
        </button>
      </div>
      <div>
        <input arial-label="Current Page" class="form-control pager-input" max="3" min="1" type="text" value="2">
      </div>
      <span class="input-group-addon">
        vong 3
      </span>
      <div class="input-group-btn pager-buttons">
        <button aria-label="Next Page" class="btn btn-default" type="button">
          <span class="glyphicon glyphicon-chevron-right"></span>
        </button>
      </div>
    </nav>
  </div>
</kiwi-pager>
`;

exports[`kiwi-pager renders empty 1`] = `<kiwi-pager page="0" total="0"></kiwi-pager>`;
